modules/fedimint-walletv2-common/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+}
+
+pub fn tweak_public_key(pk: &PublicKey, tweak: &sha256::Hash) -> PublicKey {
+    pk.add_exp_tweak(
+        secp256k1::SECP256K1,
+        &Scalar::from_be_bytes(tweak.to_byte_array()).expect("Hash is within field order"),
+    )
+    .expect("Failed to tweak bitcoin public key")
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable, Serialize, Deserialize)]
+pub enum DestinationScript {
+    P2PKH(hash160::Hash),
+    P2SH(hash160::Hash),
+    P2WPKH(hash160::Hash),
+    P2WSH(sha256::Hash),
+    P2TR(XOnlyPublicKey),
+}
+
+impl DestinationScript {
+    pub fn from_address(address: &Address) -> Self {
+        if let Some(hash) = address.pubkey_hash() {
+            return DestinationScript::P2PKH(hash.to_raw_hash());
+        }
+
+        if let Some(hash) = address.script_hash() {
+            return DestinationScript::P2PKH(hash.to_raw_hash());
+        }
+
+        if let Some(program) = address.witness_program() {
+            if program.is_p2wpkh() {
+                return DestinationScript::P2WPKH(
+                    hash160::Hash::from_slice(program.program().as_bytes())
+                        .expect("Witness program is 20 bytes"),
+                );
+            }
+
+            if program.is_p2wsh() {
+                return DestinationScript::P2WSH(
+                    sha256::Hash::from_slice(program.program().as_bytes())
+                        .expect("Witness program is 32 bytes"),
+                );
+            }
+
+            if program.is_p2tr() {
+                return DestinationScript::P2TR(
+                    XOnlyPublicKey::from_slice(program.program().as_bytes())
+                        .expect("Witness program is 32 bytes"),
+                );
+            }
+        }
+
+        panic!("Failed to obtain Destination from address");
+    }
+
+    pub fn script_pubkey(&self) -> ScriptBuf {
+        match self {
+            Self::P2PKH(hash) => ScriptBuf::new_p2pkh(&PubkeyHash::from_raw_hash(*hash)),
+            Self::P2SH(hash) => ScriptBuf::new_p2sh(&ScriptHash::from_raw_hash(*hash)),
+            Self::P2WPKH(hash) => ScriptBuf::new_p2wpkh(&WPubkeyHash::from_raw_hash(*hash)),
+            Self::P2WSH(hash) => ScriptBuf::new_p2wsh(&WScriptHash::from_raw_hash(*hash)),
+            Self::P2TR(pk) => ScriptBuf::new_p2tr_tweaked(pk.dangerous_assume_tweaked()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Encodable, Decodable)]
+pub struct FederationWallet {
+    pub value: bitcoin::Amount,
+    pub outpoint: bitcoin::OutPoint,
+    pub tweak: sha256::Hash,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, Encodable, Decodable)]
+pub struct SendFee {
+    // This index ensures that once a Fedimint transaction becomes invalid due to a
+    // change in feerate it stays invalid. Otherwise there exists a race condition
+    // where a client might consider a actually accepted transaction as rejected.
```

Comment:
```suggestion
    // change in feerate, it remains invalid. Otherwise, there exists a race condition
    // where a client might consider an actually accepted transaction as rejected.
```